# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [23]:
import sklearn
import pandas as pd
from sklearn import datasets, svm, metrics
from sklearn.linear_model import LogisticRegression
import pandas as pd
from sklearn.metrics import classification_report
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_diabetes

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [33]:
diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [32]:
diabetes.keys()


dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [34]:
# Your code here:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, T-Cells (a type of white blood cells)
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, thyroid stimulating hormone
      - s5      ltg, lamotrigine
      - s6      glu, blood sugar level

Note: Each of these 10 feature va

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

* Tenemos un total de 10 atributos: age, sex, body mass index, average blood pressure, and six blood serum measurements
* Data series la utilizamos para entrenar el modelo, y target series para testearlo.
* Tenemos un total de 442 muestras.

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [59]:
#print(len(diabetes["data"]))
#print(len(diabetes["target"]))
print(diabetes["data"].shape)
print(diabetes["target"].shape)
# ok

(442, 10)
(442,)


# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [49]:
from sklearn import linear_model


#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [50]:
diabetes_model = linear_model.LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [58]:
diabetes_data_train = diabetes['data'][:-20]
diabetes_target_train = diabetes['target'][:-20]
diabetes_data_test = diabetes['data'][-20:]
diabetes_target_test = diabetes['target'][-20:]
#diabetes_data_train, diabetes_target_train, diabetes_data_test, diabetes_target_test = train_test_split(X, y, test_size=0.2, random_state=0)


Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [62]:
fit_model = diabetes_model.fit(diabetes_data_train,diabetes_target_train)

#print(f'Coefficients: {fit_model.coef_}')
#print(f'Intercept : {fit_model.intercept_}')
print(fit_model.intercept_)
print(fit_model.coef_)

#Un estadístico podría meter su cabeza en un horno y sus pies en hielo y decir que en promedio se encuentra bien.

152.76430691633442
[ 3.03499549e-01 -2.37639315e+02  5.10530605e+02  3.27736980e+02
 -8.14131709e+02  4.92814588e+02  1.02848452e+02  1.84606489e+02
  7.43519617e+02  7.60951722e+01]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [65]:
Pedict = diabetes_model.predict(diabetes_data_test)
print(Pedict)

[197.61846908 155.43979328 172.88665147 111.53537279 164.80054784
 131.06954875 259.12237761 100.47935157 117.0601052  124.30503555
 218.36632793  61.19831284 132.25046751 120.3332925   52.54458691
 194.03798088 102.57139702 123.56604987 211.0346317   52.60335674]


#### Print your `diabetes_target_test` and compare with the prediction. 

In [66]:
pd.DataFrame({'target': diabetes_target_test, 'predicted': Pedict})

target   predicted
0    233.0  197.618469
1     91.0  155.439793
2    111.0  172.886651
3    152.0  111.535373
4    120.0  164.800548
5     67.0  131.069549
6    310.0  259.122378
7     94.0  100.479352
8    183.0  117.060105
9     66.0  124.305036
10   173.0  218.366328
11    72.0   61.198313
12    49.0  132.250468
13    64.0  120.333292
14    48.0   52.544587
15   178.0  194.037981
16   104.0  102.571397
17   132.0  123.566050
18   220.0  211.034632
19    57.0   52.603357

#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [71]:
# Your explanation here:
score = diabetes_model.score(diabetes_data_test,diabetes_target_test)
print(f"El modelo ha predecido un total de {round(score*100)}%")

El modelo ha predecido un total de 59.0%


# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [ ]:
#Venga vengaaa que hoy estoy motivado bonus y todo!

#Bueno mejor no jajajaj esq he sacado el de abajo y me he venido arriba!

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

In [ ]:
#lo dicho jajaja


# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [72]:
auto = pd.read_csv('../auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [73]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [78]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [86]:
# Your code here:
display(auto.sort_values(by='model_year', ascending=False).head(1))
display(auto.sort_values(by='model_year').head(1))

mpg  cylinders  displacement  horse_power  weight  acceleration  \
397  31.0          4         119.0         82.0    2720          19.4   

     model_year        car_name  
397          82  \t"chevy s-10"

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"

Check the dataset for missing values and remove all rows containing at least one missing value.

In [93]:
# Your code here:
print(auto.isnull().sum())
auto.dropna(inplace=True)
print(auto.isnull().sum())

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64
mpg             0
cylinders       0
displacement    0
horse_power     0
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64


Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [94]:
# Your code here:
auto["cylinders"].value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [97]:
auto.drop(['car_name'], axis=1,inplace=True)

In [98]:
auto

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0    18.0          8         307.0        130.0    3504          12.0   
1    15.0          8         350.0        165.0    3693          11.5   
2    18.0          8         318.0        150.0    3436          11.0   
3    16.0          8         304.0        150.0    3433          12.0   
4    17.0          8         302.0        140.0    3449          10.5   
..    ...        ...           ...          ...     ...           ...   
393  27.0          4         140.0         86.0    2790          15.6   
394  44.0          4          97.0         52.0    2130          24.6   
395  32.0          4         135.0         84.0    2295          11.6   
396  28.0          4         120.0         79.0    2625          18.6   
397  31.0          4         119.0         82.0    2720          19.4   

     model_year  
0            70  
1            70  
2            70  
3            70  
4            70  
..          ...  
393          82  
394          82  
395          82  
396          82  
397          82  

[392 rows x 7 columns]

In [104]:
y= auto["mpg"]

In [105]:
x= auto.drop(['mpg'], axis=1)

In [107]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2, random_state=1)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [110]:
# Your code here:
auto_model = linear_model.LinearRegression()

In [114]:
auto_fit_model = auto_model.fit(x_train,y_train)

print(auto_fit_model.intercept_)
print(auto_fit_model.coef_)

-16.372523145622143
[ 0.05210727  0.00536766  0.00728024 -0.00735996  0.1721894   0.74830698]


In [123]:
Pedict = auto_fit_model.predict(x_test)
print(Pedict)

[23.11699122 27.80216223 31.17136662 21.48775234 34.27965452 30.05172326
 25.72009165 12.78228199 24.99207912 30.8718869  23.90216766 31.53121782
 20.13361512 30.99906612 30.70437769 15.48956906 31.32860371 31.42838151
  9.96882278 23.34456511 26.25946487 17.86854534 18.34726786 16.08901013
  9.94233483 19.4251906  30.61101098 18.92372789 22.3198151  30.45267835
 22.17007958 25.49545437 10.11956193 29.02766769 21.26624211 14.78825599
 17.37663092 21.6009888  34.29382613 24.19874481 15.46315561 11.13638146
 21.95721646 22.69193764 22.59575709 18.06776537  7.54276406 34.71199899
 21.60415669  9.64242823 15.60063845 23.98638526 28.29734819 31.63953576
 33.68492799 22.22181539 27.66884238 21.72876917 27.87873164 32.70509553
 28.06337347 26.44166471 34.18278641 22.5113597  23.03701911 21.25804318
 21.69588701 14.23107755 32.18949992  6.05195193 29.67769022 21.55395817
 18.48861027 31.40598061 23.39975981 32.44580644 19.51521378 11.1212454
 11.61215173]


In [125]:
pd.DataFrame({'target': y_test, 'predicted': Pedict})
#oh mama

target  predicted
82     23.0  23.116991
167    29.0  27.802162
356    32.4  31.171367
120    19.0  21.487752
385    38.0  34.279655
..      ...        ...
23     26.0  23.399760
295    35.7  32.445806
13     14.0  19.515214
91     13.0  11.121245
62     13.0  11.612152

[79 rows x 2 columns]

In [130]:
auto_score = auto_fit_model.score(x_test,y_test)
print(f"El modelo ha predecido un total de {round(auto_score*100)}%")
#lo he petao eh! 
#dime con quien te juntas y te dire quien ereee! con un 81% de fiabilidad. :)!

El modelo ha predecido un total de 81.0%


# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [128]:
from sklearn.metrics import r2_score

y_pred = auto_model.predict(x_train)

r2 = r2_score(y_train,y_pred)
print(f'The r^2 score is {r2*100}%')

The r^2 score is 80.5894855213757%


#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [131]:
y_test_pred = auto_model.predict(x_test)

r2 = r2_score(y_test,y_test_pred)
print(f'The r^2 score is {r2*100}%')

The r^2 score is 81.44902542572598%


#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [132]:
x_train09,x_test09,y_train09,y_test09 = train_test_split(x,y,test_size=0.1)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [133]:
auto_model09=linear_model.LinearRegression()


Compute the predicted values and r squared score for our new model and new sample data.

In [137]:
y_pred = auto_model.predict(x_train09)

r2 = r2_score(y_train09,y_pred)
print(f'The r^2 score is {r2*100}%')

The r^2 score is 80.77835559813094%


Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [138]:
y_test_pred = auto_model.predict(x_test09)

r2 = r2_score(y_test09,y_test_pred)
print(f'The r^2 score is {r2*100}%')

The r^2 score is 80.6865524411224%


# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [ ]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [146]:
from sklearn.datasets import make_friedman1
from sklearn.feature_selection import RFE
from sklearn.svm import SVR
estimator = SVR(kernel="linear")
selector = RFE(estimator, n_features_to_select=3, step=1)
selector = selector.fit(x_train, y_train)


In [148]:
selector.support_

array([ True, False, False, False,  True,  True])

Fit the model and print the ranking

In [164]:
print(selector.ranking_)
print(x_train.columns)

[1 3 2 4 1 1]
Index(['cylinders', 'displacement', 'horse_power', 'weight', 'acceleration',
       'model_year'],
      dtype='object')


Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [190]:
auto_reduced=auto[["cylinders","acceleration","model_year"]]#No entiendo la logica detras de poner doble corchete...


y=auto["mpg"]
X=auto_reduced


X_train_reduced,X_test_reduced,y_train_reduced,y_test_reduced= train_test_split(X,y,test_size=0.2)


In [191]:
auto_model_reduced = linear_model.LinearRegression()
auto_model_reduced.fit(X_train_reduced,y_train_reduced)


LinearRegression()

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [192]:
y_pred_reduced = auto_model_reduced.predict(X_train_reduced)


r2 = r2_score(y_train_reduced,y_pred_reduced)
print(f'The r^2 score is {r2*100}%')

#Se ve que o la he cagado o r2 baja... a ver es normal menos columnas pues menos capacidad de prediccion a no seeer que las columnas molesten mas que ayuden. 


The r^2 score is 72.14926454680237%


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.